### PR TITLE
planner, executor: fix error code

### PR DESCRIPTION
**Pull request by ichn-hu**
### What problem does this PR solve? 

Associating internal error code for planner/core/errors.go, expression/errors.go and executor/errors.go to parser/mysql/errcode.go.

### What is changed and how it works?

Adding more error code and set up the mapping.

### Check List 

Tests 

 - No code

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

File: planner/core/errors.go [hunk 1]
```diff
@@ -19,74 +19,73 @@ import (
...
-	codeWindowFrameEndIllegal           = mysql.ErrWindowFrameEndIllegal
-	codeWindowFrameIllegal              = mysql.ErrWindowFrameIllegal
-	codeWindowRangeFrameOrderType       = mysql.ErrWindowRangeFrameOrderType
-	codeWindowRangeFrameTemporalType    = mysql.ErrWindowRangeFrameTemporalType
-	codeWindowRangeFrameNumericType     = mysql.ErrWindowRangeFrameNumericType
-	codeWindowRangeBoundNotConstant     = mysql.ErrWindowRangeBoundNotConstant
-	codeWindowRowsIntervalUse           = mysql.ErrWindowRowsIntervalUse
-	codeWindowFunctionIgnoresFrame      = mysql.ErrWindowFunctionIgnoresFrame
-	codeUnsupportedOnGeneratedColumn    = mysql.ErrUnsupportedOnGeneratedColumn
+	codeUnsupportedType                 terror.ErrCode = mysql.ErrUnsupportedType
```
Comment (jackysp) [inline planner/core/errors.go]: Could we avoid this part?

File: planner/core/errors.go [hunk 2]
```diff
@@ -19,74 +19,73 @@ import (
...
-	ErrWrongParamCount             = terror.ClassOptimizer.New(codeWrongParamCount, "Wrong parameter count")
-	ErrSchemaChanged               = terror.ClassOptimizer.New(codeSchemaChanged, "Schema has changed")
+	ErrUnsupportedType             = terror.ClassOptimizer.New(codeUnsupportedType, mysql.MySQLErrName[mysql.ErrUnsupportedType])
+	ErrAnalyzeMissIndex            = terror.ClassOptimizer.New(codeAnalyzeMissIndex, mysql.MySQLErrName[mysql.ErrAnalyzeMissIndex])
+	ErrCartesianProductUnsupported = terror.ClassOptimizer.New(codeUnsupported, mysql.MySQLErrName[mysql.ErrCartesianProductUnsupported])
+	ErrStmtNotFound                = terror.ClassOptimizer.New(codeStmtNotFound, mysql.MySQLErrName[mysql.ErrPreparedStmtNotFound])
+	ErrWrongParamCount             = terror.ClassOptimizer.New(codeWrongParamCount, mysql.MySQLErrName[mysql.ErrWrongParamCount])
+	ErrSchemaChanged               = terror.ClassOptimizer.New(codeSchemaChanged, mysql.MySQLErrName[mysql.ErrSchemaChanged])
 	ErrTablenameNotAllowedHere     = terror.ClassOptimizer.New(codeTablenameNotAllowedHere, "Table '%s' from one of the %ss cannot be used in %s")
```
Comment (jackysp) [inline planner/core/errors.go]: We should use the predefined error message.

File: planner/core/errors.go [hunk 3]
```diff
@@ -147,6 +146,12 @@ var (
 
 func init() {
 	mysqlErrCodeMap := map[terror.ErrCode]uint16{
+		codeUnsupportedType:              mysql.ErrUnsupportedType,
```
Comment (jackysp) [inline planner/core/errors.go]: Please add some tests for this map.

File: planner/core/errors.go (hunk 2 above)
Comment (ichn-hu) [reply planner/core/errors.go]: This one is a little different, it has a predefined error message in `mysql/errname.go` for `ErrTablenameNotAllowedHere`, but it is different from the native error message.

```
	ErrTablenameNotAllowedHere:                  "Table '%-.192s' from one of the SELECTs cannot be used in %-.32s",
```

One work around is to then create another error code for this error, say `ErrTablenameNotAllowedHere2` if we want every error message to be strictly defined in `mysql/errname.go`.

File: planner/core/errors.go (hunk 3 above)
Comment (ichn-hu) [reply planner/core/errors.go]: Could you give some hint on how to add tests for this map?

File: planner/core/errors.go (hunk 1 above)
Comment (ichn-hu) [reply planner/core/errors.go]: Sorry I didn't get what you mean? Do you mean we want to totally get ride of `codeXXX` and just use `mysql.ErrXX`?

Comment (jackysp) [reply planner/core/errors.go]: yes

File: planner/core/errors.go (hunk 2 above)
Comment (jackysp) [reply planner/core/errors.go]: I think we could modify the error message in  mysql/errname.go

File: planner/core/errors.go (hunk 3 above)
Comment (jackysp) [reply planner/core/errors.go]: maybe like https://github.com/pingcap/tidb/pull/13300/files#diff-8715d6cdb7a45a830b51e3bb85466eeaR26

**Comment by ichn-hu** [comment]
@jackysp PTAL again.

**Comment by jackysp** [comment]
LGTM, please fix the build and test.

**Comment by ichn-hu** [comment]
Will do, https://github.com/pingcap/parser/pull/650 merged.

**Comment by ichn-hu** [comment]
@jackysp PTAL again

**Comment by ichn-hu** [comment]
@jackysp PR https://github.com/pingcap/tidb-test/pull/961 fixes common-test, both PR are ready for merge, PTAL.

**Pull request merged**